engine/consensus/dkg/reactor_engine.go

```diff
@@ -161,34 +229,90 @@ func (e *ReactorEngine) EpochSetupPhaseStarted(currentEpochCounter uint64, first
+	// check previously.
+	endState, err := e.dkgState.GetDKGEndState(nextEpochCounter)
+	if err == nil {
+		log.Warn().Msgf("checking beacon key consistency: exiting because dkg end state was already set: %s", endState.String())
 		return
 	}
 
-	dkgPrivInfo, err := e.keyStorage.RetrieveMyBeaconPrivateKey(currentEpochCounter + 1)
+	// Since epoch phase transitions are emitted when the first block of the new
+	// phase is finalized, the block's snapshot is guaranteed to already be
+	// accessible in the protocol state at this point (even though the Badger
+	// transaction finalizing the block has not been committed yet).
+	nextDKG, err := e.State.AtBlockID(firstBlock.ID()).Epochs().Next().DKG()
 	if err != nil {
-		e.log.Err(err).Msg("checking DKG key consistency: could not retrieve DKG private info for next epoch")
+		// CAUTION: this should never happen, indicates a storage failure or corruption
+		log.Fatal().Err(err).Msg("checking beacon key consistency: could not retrieve next DKG info")
+		return
+	}
+
+	myBeaconPrivKey, err := e.dkgState.RetrieveMyBeaconPrivateKey(nextEpochCounter)
+	if errors.Is(err, storage.ErrNotFound) {
+		log.Warn().Msg("checking beacon key consistency: no key found")
+		err := e.dkgState.SetDKGEndState(nextEpochCounter, flow.DKGEndStateNoKey)
+		if err != nil {
+			log.Fatal().Err(err).Msg("failed to set dkg end state")
+		}
+		return
+	} else if err != nil {
+		log.Fatal().Err(err).Msg("checking beacon key consistency: could not retrieve beacon private key for next epoch")
 		return
 	}
 
 	nextDKGPubKey, err := nextDKG.KeyShare(e.me.NodeID())
 	if err != nil {
-		e.log.Err(err).Msg("checking DKG key consistency: could not retrieve DKG public key for next epoch")
+		log.Fatal().Err(err).Msg("checking beacon key consistency: could not retrieve my beacon public key for next epoch")
 		return
 	}
+	localPubKey := myBeaconPrivKey.PublicKey()
 
-	localPubKey := dkgPrivInfo.PublicKey()
-
+	// we computed a local beacon key but it is inconsistent with our canonical
+	// public key - therefore it is unsafe for use
 	if !nextDKGPubKey.Equals(localPubKey) {
-		e.log.Warn().Msg("checking DKG key consistency: locally computed dkg public key does not match dkg public key for next epoch")
+		log.Warn().
+			Hex("computed_beacon_pub_key", localPubKey.Encode()).
+			Hex("canonical_beacon_pub_key", nextDKGPubKey.Encode()).
+			Msg("checking beacon key consistency: locally computed beacon public key does not match beacon public key for next epoch")
+		err := e.dkgState.SetDKGEndState(nextEpochCounter, flow.DKGEndStateInconsistentKey)
+		if err != nil {
+			log.Fatal().Err(err).Msg("failed to set dkg end state")
+		}
+		return
 	}
+
+	err = e.dkgState.SetDKGEndState(nextEpochCounter, flow.DKGEndStateSuccess)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("failed to set dkg")
+	}
+	log.Info().Msgf("successfully ended DKG, my beacon pub key for epoch %d is %x", nextEpochCounter, localPubKey.Encode())
```

Comment:
I guess this should work
```suggestion
	log.Info().Msgf("successfully ended DKG, my beacon pub key for epoch %d is %s", nextEpochCounter, localPubKey)
```